File: 2626-count-the-number-of-good-subarrays/count-the-number-of-good-subarrays.go
func countGood(nums []int, k int) int64 {
    if len(nums) < 2 {
        return 0;
    }
    
    freq := make(map[int]int);
    n := len(nums);
    left := 0;
    var pairs, result int64 = 0, 0;

    for right := 0; right < n; right++ {
        val := nums[right];
        count := freq[val];
        pairs += int64(count);
        freq[val]++;

        for pairs >= int64(k) {
            result += int64(n - right);
            leftVal := nums[left];
            freq[leftVal]--;
            pairs -= int64(freq[leftVal]);
            left++;
        }
    }

    return result;
}

File: 2626-count-the-number-of-good-subarrays/count-the-number-of-good-subarrays.java
class Solution {
    public long countGood(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        int n = nums.length, left = 0;
        long pairs = 0, result = 0;

        for (int right = 0; right < n; right++) {
            int val = nums[right];
            int count = freq.getOrDefault(val, 0);
            
            pairs += count;
            freq.put(val, count + 1);

            while (pairs >= k) {
                result += n - right;

                int leftVal = nums[left];
                int leftCount = freq.get(leftVal);

                pairs -= leftCount - 1;
                freq.put(leftVal, leftCount - 1);
                left++;
            }
        }

        return result;
    }
}